File: typescript/src/core/modules/auth/functions/sign-in.ts
import { InitiateAuthCommand } from "@aws-sdk/client-cognito-identity-provider";

// Import constants
import { Configs } from "../../../../utils/configs";

// Import errors
import { AppError, isStandardError } from "../../../error";

// Import aws clients from utils
import { getCognitoIDProviderClient } from "../../../../utils/aws-clients";

// Import types
import type { InitiateAuthCommandInput } from "@aws-sdk/client-cognito-identity-provider";
import type { TRuntimeContext } from "../../../context/runtime-context";

/**
 * Cho phép một người dùng đăng nhập vào trong hệ thống.
 *
 * @param ctx - runtime context
 *
 * @returns
 */
export async function signIn(ctx: TRuntimeContext) {
  try {
    const body = await ctx.getBody<{ username: string; password: string }>();

    const input: InitiateAuthCommandInput = {
      AuthFlow: "USER_PASSWORD_AUTH",
      AuthParameters: {
        PASSWORD: body.password,
        USERNAME: body.username,
      },
      ClientId: Configs.CognitoAppClientId,
    };

    const client = getCognitoIDProviderClient({});
    const command = new InitiateAuthCommand(input);
    const response = await client.send(command);

    const authenticationResult = response.AuthenticationResult!;

    return {
      auth: {
        idToken: authenticationResult.IdToken,
        accessToken: authenticationResult.AccessToken,
        refreshToken: authenticationResult.RefreshToken,
        expiresIn: authenticationResult.ExpiresIn,
      },
    };
  } catch (error: any) {
    if (isStandardError(error)) return error;

    const err = new AppError("Cannot authenticate user");
    err.asHTTPError("InternalServerError");
    err.addErrorDetail({ source: "signIn", desc: error.message });
    return err;
  }
}

File: python/src/core/modules/auth/functions/sign_in.py
# Import from core
from core.error.main import AppError, is_standard_error

# Import from utils
from utils.aws_clients.main import get_cognito_idp_client
from utils.configs import Configs


async def sign_in(ctx):
    """
    Cho phép một người dùng đăng nhập vào trong hệ thống.

    Args:
        ctx: runtime context

    Returns:
        dict: chứa tokens hoặc AppError
    """
    try:
        body = await ctx.get_body()
        username = body.get("username")
        password = body.get("password")

        print("Body:", body)
        print("Path Params:", ctx.get_params())
        print("Query:", ctx.get_query())

        client = get_cognito_idp_client()

        response = client.initiate_auth(
            AuthFlow="USER_PASSWORD_AUTH",
            AuthParameters={
                "USERNAME": username,
                "PASSWORD": password,
            },
            ClientId=Configs.Cognito_App_Client_Id,
        )

        auth_result = response.get("AuthenticationResult", {})

        return {
            "auth": {
                "idToken": auth_result.get("IdToken"),
                "accessToken": auth_result.get("AccessToken"),
                "refreshToken": auth_result.get("RefreshToken"),
                "expiresIn": auth_result.get("ExpiresIn"),
            }
        }

    except Exception as error:
        if is_standard_error(error):
            return error

        err = AppError("Cannot authenticate user")
        err.as_http_error("InternalServerError")
        err.add_error_detail({"source": "signIn", "desc": str(error)})

        return err
